src/Stratis.Bitcoin.Features.BlockStore/PrunedBlockRepository.cs

```diff
@@ -0,0 +1,179 @@
+        private readonly ILogger logger;
+        private static readonly byte[] prunedTipKey = new byte[2];
+        private readonly StoreSettings storeSettings;
+
+        /// <inheritdoc />
+        public HashHeightPair PrunedTip { get; private set; }
+
+        public PrunedBlockRepository(IBlockRepository blockRepository, ILoggerFactory loggerFactory, StoreSettings storeSettings)
+        {
+            this.blockRepository = blockRepository;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.storeSettings = storeSettings;
+        }
+
+        /// <inheritdoc />
+        public Task InitializeAsync(Network network)
+        {
+            Block genesis = network.GetGenesis();
+
+            Task task = Task.Run(() =>
+            {
+                using (DBreeze.Transactions.Transaction transaction = this.blockRepository.DBreeze.GetTransaction())
+                {
+                    bool doCommit = false;
+
+                    if (this.storeSettings.Prune)
+                    {
+                        if (this.LoadPrunedTip(transaction) == null)
+                        {
+                            this.PrunedTip = new HashHeightPair(genesis.GetHash(), 0);
+                            transaction.Insert(BlockRepository.CommonTableName, prunedTipKey, this.PrunedTip);
+                            doCommit = true;
+                        }
+                    }
+
+                    if (doCommit) transaction.Commit();
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public async Task PruneDatabase(ChainedHeader consensusTip, bool nodeInitializing)
+        {
+            this.logger.LogInformation($"Pruning started.");
+
+            if (nodeInitializing)
+            {
+                if (IsDatabasePruned())
+                    return;
+
+                await this.PrepareDatabaseForPruningAsync(consensusTip);
+            }
+
+            this.CompactDataBase();
+
+            this.logger.LogInformation($"Pruning complete.");
+
+            return;
+        }
+
+        private bool IsDatabasePruned()
+        {
+            if (this.blockRepository.TipHashAndHeight.Height <= this.PrunedTip.Height + this.storeSettings.PruneBlockMargin)
+            {
+                this.logger.LogDebug("(-):true");
+                return true;
+            }
+            else
+            {
+                this.logger.LogDebug("(-):false");
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Compacts the block and transaction database by recreating the tables without the deleted references.
+        /// </summary>
+        /// <param name="consensusTip">The last fully validated block of the node.</param>
+        /// <returns>The awaited task.</returns>
+        private async Task PrepareDatabaseForPruningAsync(ChainedHeader consensusTip)
```

Comment:
The name consensusTip is not correct here right? it should be storeTip

Reply:
Why the name `PrepareDatabaseForPruningAsync` this is actually the pruning code no?

Reply:
it should be repo tip. 
in case store is > prune amount ahead of the the repo (which is almost always the case in IBD) and then we kill the node, start it and have a reorg- then the node is dead and can't recover.

Reply:
> 
> 
> The name consensusTip is not correct here right? it should be storeTip

Yes it should be the repository's tip.... Fixed!

Reply:
> 
> 
> Why the name `PrepareDatabaseForPruningAsync` this is actually the pruning code no?

Yes this should actually be called `PrepareDatabaseForCompactingAsync`... Fixed!

Reply:
>it should be repo tip.
`StoreTip` and `RepoTip` are the same thing I believe, the cache has its own tip `StoreCacheTip`